k2/csrc/cuda/log.h

```diff
@@ -0,0 +1,185 @@
+// do not follow the google c++ code style
+constexpr LogLevel DEBUG = LogLevel::kDebug;
+constexpr LogLevel INFO = LogLevel::kInfo;
+constexpr LogLevel WARNING = LogLevel::kWarning;
+constexpr LogLevel ERROR = LogLevel::kError;
+constexpr LogLevel FATAL = LogLevel::kFatal;
+
+class Logger {
+ public:
+  __host__ __device__ Logger(const char *filename, const char *func_name,
+                             uint32_t line_num, LogLevel level)
+      : filename_(filename),
+        func_name_(func_name),
+        line_num_(line_num),
+        level_(level) {
+    switch (level) {
+      case DEBUG:
+        printf("[D] ");
+        break;
+      case INFO:
+        printf("[I] ");
+        break;
+      case WARNING:
+        printf("[W] ");
+        break;
+      case ERROR:
+        printf("[E] ");
+        break;
+      case FATAL:
+        printf("[F] ");
+        break;
+    }
+    printf("%s:%s:%u ", filename, func_name, line_num);
+#if defined(__CUDA_ARCH__)
+    printf("block:[%u,%u,%u], thread: [%u,%u,%u] ", blockIdx.x, blockIdx.y,
+           blockIdx.z, threadIdx.x, threadIdx.y, threadIdx.z);
+#endif
+  }
+
+  __host__ __device__ ~Logger() {
+    printf("\n");
+    if (level_ == FATAL) {
+#if defined(__CUDA_ARCH__)
+      // this is usually caused by one of the CHECK macros and the detailed
+      // error messages should have already been printed by the macro, so we
+      // use an arbitrary string here.
+      __assert_fail("Some bad things happened", filename_, line_num_,
+                    func_name_);
+#else
+      abort();
+#endif
+    }
+  }
+
+  __host__ __device__ const Logger &operator<<(const char *s) const {
+    printf("%s", s);
+    return *this;
+  }
+
+  __host__ __device__ const Logger &operator<<(int32_t i) const {
+    printf("%d", i);
+    return *this;
+  }
+
+  __host__ __device__ const Logger &operator<<(uint32_t i) const {
+    printf("%u", i);
+    return *this;
+  }
+
+  __host__ __device__ const Logger &operator<<(double d) const {
+    printf("%f", d);
+    return *this;
+  }
+
+ private:
+  const char *filename_;
+  const char *func_name_;
+  uint32_t line_num_;
+  LogLevel level_;
+};
+
+class Voidifier {
+ public:
+  __host__ __device__ void operator&(const Logger &) const {}
+};
+
+}  // namespace internal
+
+}  // namespace k2
+
+#define K2_CHECK(x)                                                           \
+  (x) ? (void)0                                                               \
+      : k2::internal::Voidifier() & k2::internal::Logger(__FILE__, __func__,  \
+                                                         __LINE__,            \
+                                                         k2::internal::FATAL) \
+                                        << "Check failed: " << #x << " "
+
+// WARNING: x and y are may be evaluated multiple times, but this happens only
+// when the check fails. Since the program aborts if it fails, we don't think
+// the extra evaluation of x and y matters.
```

Comment:
I also think it doesn't matter to evaluate twice, but maybe it's better to add comments to reminds users that the program will get some undefined behaviors if `x` or `y` is a function which has side effect when called, e.g. they will change some static (global) variables, even it's unlikely to happen with our own code.

Reply:
Done.